src/language/syntax/rule.ts

```diff
@@ -174,19 +186,26 @@ export abstract class SyntaxRule<T> {
 
   // Combinators
 
-  or<R>(rule: SyntaxRule<R>): SyntaxRuleOr<T, R> {
+  static or<R extends SyntaxRule<unknown>[]>(
+    ...rules: R
+  ): SyntaxRuleOr<PeelTupleSyntaxRule<R>> {
+    return new SyntaxRuleOr(...rules) as SyntaxRuleOr<PeelTupleSyntaxRule<R>>;
```

Comment:
The `as` here is weird because I don't really understand why it's needed. My expectation is that `PeelTupleSyntaxRule<R extends SyntaxRule<unknown>[]>` returns `T extends unknown[]` while preserving the provenance of the type back to `R`.

However, that doesn't happen for some reasaon. This mean that `SyntaxRuleOr(...rules)` returns `SyntaxRuleOr<unknown[]>` instead of expected `SyntaxRuleOr<T extends unknown[]>` (where `T` comes from peeling `R` 🍌, i.e. the provenance).

As for assert.. I guess we could assert it at runtime, when parsing, but I don't like that because we are losing runtime performance for something we can just test thoroughly as dev time. We cannot assert at rule construction time because we'd have to assert a function implementation returns something and that's not any more possible for us than for the compiler, and the compiler is apparently struggling with it already.

I hate it too.

Reply:
Weird indeed. Good point about performance though. 👍🏻

Reply:
Keep it then.

Reply:
you kids and your performance. *laughs in javascript*